# Data Collection

Data collection is the systematic process of gathering information from diverse sources, including databases, surveys, and sensors. It is a critical step in obtaining the raw material for analysis and decision-making. **Ethical considerations** are fundamental in this process, emphasizing principles such as informed consent, privacy protection, and data security. Adhering to ethical standards is essential for building trust and ensuring fairness in the treatment of individuals whose data is being collected.
Furthermore, **data scrubbing**, or cleaning, is an integral part of data collection, involving the correction of errors and handling missing values. This process enhances the quality and reliability of the collected data, ensuring that it is consistent and ready for effective analysis without introducing biases or inaccuracies.

<img src= "https://uploads-ssl.webflow.com/5a00e7aa079aa40001b3c4fb/5d5c22e040c6beab16860e8e_data-cleaning-thumb.png" >

We will leverage two datasets that provide information on crime in Houston. The first dataset spans from 2009 to 2018, while the second one covers the period from 2019 to 2021. Our objective is to consolidate these datasets into a unified and cleaned dataset for comprehensive analysis. During the data cleaning process, we will focus on extracting relevant information, including the neighborhood where each crime occurred, the corresponding year, and the type of crime. Furthermore, based on the type of crime, we will categorize offenses into serious and non-serious categories. This approach ensures that our consolidated dataset is tailored to our specific analytical needs, allowing us to gain insights into crime trends in Houston over the entire period.

This data was taken from The Kinder Institute. Here is the link where you can find the dataset and more:
https://www.kinderudp.org/#/datasetCatalog/g2lj2kjrv487

In [56]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [57]:
# Importing the pandas library and reading the data from a CSV file
crimen = pd.read_csv('vi_hpd_crime_CTA_nibrsCode_TabDelim.txt', sep='\t')

# Displaying the contents of the 'crimen' DataFrame
crimen

2019  NULL   NULL.1  09A Murder & Nonnegligent Manslaughter  A  \
0      2019   NaN      NaN  09B             Negligent Manslaughter  A   
1      2019   NaN      NaN  100               Kidnapping/Abduction  A   
2      2019   NaN      NaN  11A                      Forcible Rape  A   
3      2019   NaN      NaN  11B                    Forcible Sodomy  A   
4      2019   NaN      NaN  11C      Sexual Assault With An Object  A   
...     ...   ...      ...  ...                                ... ..   
13675  2021  K140  Baytown  90Z                 All Other Offenses  B   
13676  2021  K143  Tomball  13B                     Simple Assault  A   
13677  2021  K143  Tomball  23D                Theft From Building  A   
13678  2021  K143  Tomball  23H                  All Other Larceny  A   
13679  2021  K143  Tomball  90Z                 All Other Offenses  B   

                             Person  14  
0                            Person   1  
1                            Person   8  
2                            Person  27  
3                            Person  16  
4                            Person   1  
...                             ...  ..  
13675  Person, Property, or Society   1  
13676                        Person   1  
13677                      Property   1  
13678                      Property   4  
13679  Person, Property, or Society   1  

[13680 rows x 8 columns]

We observed that there are some missing values in our dataset, specifically in the 'NULL' column. However, given that there are only 40 missing values out of a total of 13,679 rows, which constitutes a negligible portion of the dataset, we have decided to eliminate these rows. The removal of these rows is justified by the insignificance of the missing data in the context of our entire time period. This ensures that our analysis is not adversely affected, and we can proceed with a more complete and accurate dataset for our comprehensive examination of crime trends.

In [58]:
# Filtering rows where the 'NULL' column is not NaN (not missing)
crimen = crimen[crimen['NULL'].isna() == False]

# Renaming columns 'NULL' and 'NULL.1' to 'cta_code' and 'cta_name', respectively
crimen.rename(columns={'NULL': 'cta_code', 'NULL.1': 'cta_name'}, inplace=True)

# Displaying the contents of the DataFrame after renaming columns
crimen

/var/folders/_n/vpsjjrqj2yb_yy6q_1xgvwg40000gn/T/ipykernel_2960/480003583.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crimen.rename(columns={'NULL': 'cta_code', 'NULL.1': 'cta_name'}, inplace=True)


2019 cta_code  cta_name  09A  Murder & Nonnegligent Manslaughter  A  \
42     2019     K001  Downtown  09A  Murder & Nonnegligent Manslaughter  A   
43     2019     K001  Downtown  11A                       Forcible Rape  A   
44     2019     K001  Downtown  11B                     Forcible Sodomy  A   
45     2019     K001  Downtown  11D                   Forcible Fondling  A   
46     2019     K001  Downtown  120                             Robbery  A   
...     ...      ...       ...  ...                                 ... ..   
13675  2021     K140   Baytown  90Z                  All Other Offenses  B   
13676  2021     K143   Tomball  13B                      Simple Assault  A   
13677  2021     K143   Tomball  23D                 Theft From Building  A   
13678  2021     K143   Tomball  23H                   All Other Larceny  A   
13679  2021     K143   Tomball  90Z                  All Other Offenses  B   

                             Person  14  
42                           Person   1  
43                           Person   2  
44                           Person   3  
45                           Person   1  
46                         Property  24  
...                             ...  ..  
13675  Person, Property, or Society   1  
13676                        Person   1  
13677                      Property   1  
13678                      Property   4  
13679  Person, Property, or Society   1  

[13534 rows x 8 columns]

In [59]:
# Renaming the column 'Murder & Nonnegligent Manslaughter' to 'crimes'
crimen.rename(columns={'Murder & Nonnegligent Manslaughter': 'crimes'}, inplace=True)

# Resetting the index of the DataFrame
crimen.reset_index(inplace=True)

# Displaying the contents of the DataFrame after renaming columns
crimen

/var/folders/_n/vpsjjrqj2yb_yy6q_1xgvwg40000gn/T/ipykernel_2960/4179663981.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crimen.rename(columns={'Murder & Nonnegligent Manslaughter': 'crimes'}, inplace=True)


index  2019 cta_code  cta_name  09A  \
0         42  2019     K001  Downtown  09A   
1         43  2019     K001  Downtown  11A   
2         44  2019     K001  Downtown  11B   
3         45  2019     K001  Downtown  11D   
4         46  2019     K001  Downtown  120   
...      ...   ...      ...       ...  ...   
13529  13675  2021     K140   Baytown  90Z   
13530  13676  2021     K143   Tomball  13B   
13531  13677  2021     K143   Tomball  23D   
13532  13678  2021     K143   Tomball  23H   
13533  13679  2021     K143   Tomball  90Z   

                                   crimes  A                        Person  14  
0      Murder & Nonnegligent Manslaughter  A                        Person   1  
1                           Forcible Rape  A                        Person   2  
2                         Forcible Sodomy  A                        Person   3  
3                       Forcible Fondling  A                        Person   1  
4                                 Robbery  A                      Property  24  
...                                   ... ..                           ...  ..  
13529                  All Other Offenses  B  Person, Property, or Society   1  
13530                      Simple Assault  A                        Person   1  
13531                 Theft From Building  A                      Property   1  
13532                   All Other Larceny  A                      Property   4  
13533                  All Other Offenses  B  Person, Property, or Society   1  

[13534 rows x 9 columns]

In [60]:
# Dropping the specified columns 'index', '09A', and 'Person' from the DataFrame
crimen.drop(['index', '09A', 'Person'], axis=1, inplace=True)

# Displaying the modified DataFrame after dropping columns
crimen

/var/folders/_n/vpsjjrqj2yb_yy6q_1xgvwg40000gn/T/ipykernel_2960/2980612632.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crimen.drop(['index', '09A', 'Person'], axis=1, inplace=True)


2019 cta_code  cta_name                              crimes  A  14
0      2019     K001  Downtown  Murder & Nonnegligent Manslaughter  A   1
1      2019     K001  Downtown                       Forcible Rape  A   2
2      2019     K001  Downtown                     Forcible Sodomy  A   3
3      2019     K001  Downtown                   Forcible Fondling  A   1
4      2019     K001  Downtown                             Robbery  A  24
...     ...      ...       ...                                 ... ..  ..
13529  2021     K140   Baytown                  All Other Offenses  B   1
13530  2021     K143   Tomball                      Simple Assault  A   1
13531  2021     K143   Tomball                 Theft From Building  A   1
13532  2021     K143   Tomball                   All Other Larceny  A   4
13533  2021     K143   Tomball                  All Other Offenses  B   1

[13534 rows x 6 columns]

In [61]:
# Renaming the columns 'crimes' to 'Offense_Type' and '14' to 'Num_Offenses'
crimen.rename(columns={'crimes': 'Offense_Type', '14': 'Num_Offenses'}, inplace=True)

# Displaying the modified DataFrame after renaming columns
crimen

/var/folders/_n/vpsjjrqj2yb_yy6q_1xgvwg40000gn/T/ipykernel_2960/2573209088.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crimen.rename(columns={'crimes': 'Offense_Type', '14': 'Num_Offenses'}, inplace=True)


2019 cta_code  cta_name                        Offense_Type  A  \
0      2019     K001  Downtown  Murder & Nonnegligent Manslaughter  A   
1      2019     K001  Downtown                       Forcible Rape  A   
2      2019     K001  Downtown                     Forcible Sodomy  A   
3      2019     K001  Downtown                   Forcible Fondling  A   
4      2019     K001  Downtown                             Robbery  A   
...     ...      ...       ...                                 ... ..   
13529  2021     K140   Baytown                  All Other Offenses  B   
13530  2021     K143   Tomball                      Simple Assault  A   
13531  2021     K143   Tomball                 Theft From Building  A   
13532  2021     K143   Tomball                   All Other Larceny  A   
13533  2021     K143   Tomball                  All Other Offenses  B   

       Num_Offenses  
0                 1  
1                 2  
2                 3  
3                 1  
4                24  
...             ...  
13529             1  
13530             1  
13531             1  
13532             4  
13533             1  

[13534 rows x 6 columns]

In [62]:
# Grouping the DataFrame by 'cta_code' and 'A', counting occurrences of '2019', and resetting the index
crimen_pt1 = crimen.groupby(['cta_code', 'A']).count()['2019'].reset_index()

# Assigning the result to a new DataFrame 'crimen_final'
crimen_final = crimen_pt1

# Displaying the contents of the 'crimen_final' DataFrame
crimen_final

cta_code  A  2019
0       K001  A   109
1       K001  B    17
2       K002  A   111
3       K002  B    14
4       K003  A   107
..       ... ..   ...
270     K140  B     6
271     K141  A    38
272     K142  A    32
273     K143  A    30
274     K143  B     1

[275 rows x 3 columns]

We have made prepared and refined our crime dataset for analysis. The initial steps involved importing the data from a CSV file, addressing missing values by filtering out rows with negligible impact, and renaming columns for clarity. Further modifications included dropping specific columns deemed irrelevant for our analysis. Subsequently, we focused on organizing the data, grouping it by 'cta_code' and 'A', and counting occurrences of crimes during the years. The results were stored in a new DataFrame named 'crimen_final'.

Now we will work on the second crime dataset from 2009 to 2018.

In [63]:
# Importing the pandas library and reading the data from a CSV file into the 'crimen_past' DataFrame
crimen_past = pd.read_csv('HPD_Crime_2009_2017_v01.csv')

# Displaying the contents of the 'crimen_past' DataFrame
crimen_past

Location_Desc Police_Beat Street_Type  RecordID  \
0                                       NaN       14D40          RD         1   
1                                       NaN        7C20         NaN         2   
2                                       NaN       13D40        PKWY         3   
3                                       NaN       18F40          RD         4   
4                      RESTAURANT/CAFETERIA       24C60        PKWY         5   
...                                     ...         ...         ...       ...   
1094202       MISC. BUSINESS (NON-SPECIFIC)        9C40          ST   1094203   
1094203               APARTMENT PARKING LOT        9C40          RD   1094204   
1094204                 SERVICE/GAS STATION        9C40          ST   1094205   
1094205  ELECTRONICS STORE, ELECTRICAL SUP.        9C40          RD   1094206   
1094206             HOTEL/MOTEL PARKING LOT         NaN         NaN   1094207   

        Date_Reported Offense_Type  Block_Range Street_Prefix   Street_Name  \
0                 NaN     Burglary    5600-5699           NaN      SELINSKY   
1                 NaN        Theft    3900-3999           NaN     CAVALCADE   
2                 NaN        Theft    8500-8599             S   SAM HOUSTON   
3                 NaN      Robbery    2700-2799           NaN       DUNVALE   
4          09/08/1914     Burglary  12700-12799             W  LAKE HOUSTON   
...               ...          ...          ...           ...           ...   
1094202    12/31/2017     Burglary  12700-12799           NaN          EAST   
1094203    12/31/2017        Theft      600-699           NaN      MAXEY RD   
1094204    12/31/2017        Theft    7600-7699           NaN          EAST   
1094205    12/31/2017      Robbery    1000-1099           NaN       FEDERAL   
1094206    12/31/2017        Theft    1300-1399           NaN   LACONCHA LN   

         Num_Offenses  ...          X          Y       GEOID10       GEOID00  \
0                   1  ... -95.333268  29.637514  4.820133e+14  4.820133e+14   
1                   1  ... -95.331799  29.800081  4.820121e+14  4.820121e+14   
2                   1  ... -95.516291  29.625560  4.820133e+14  4.820133e+14   
3                   1  ... -95.514200  29.736820  4.820143e+14  4.820143e+14   
4                   1  ... -95.195466  29.924672  4.820125e+14  4.820125e+14   
...               ...  ...        ...        ...           ...           ...   
1094202             1  ... -95.208459  29.770717  4.820123e+14  4.820123e+14   
1094203             1  ... -95.218625  29.780813  4.820123e+14  4.820123e+14   
1094204             1  ... -95.290745  29.777468  4.820121e+14  4.820121e+14   
1094205             1  ... -95.215481  29.772798  4.820123e+14  4.820123e+14   
1094206             1  ... -95.410353  29.690881  4.820131e+14  4.820131e+14   

            Police_Division  CTA_ID                      CTA_Name Geocoded  \
0        Southeast Division    K066                      Minnetex        1   
1        Northeast Division    K004              Kashmere Gardens        1   
2        Southwest Division    K070                  Five Corners        1   
3          Midwest Division    K080                      Mid-West        1   
4         Kingwood Division    K106                  Lake Houston        1   
...                     ...     ...                           ...      ...   
1094202  Northeast Division    K054                    Northshore        1   
1094203  Northeast Division    K054                    Northshore        1   
1094204  Northeast Division    K005  Denver Harbor / Port Houston        1   
1094205  Northeast Division    K054                    Northshore        1   
1094206  Southwest Division    K019                     Astrodome        1   

        Location_Code Hour  
0                 NaN    8  
1                 NaN   11  
2                 NaN   15  
3                 NaN   14  
4                 210    7  
...               ...  ...  
1094202

In [64]:
# Displaying the column names in the 'crimen_past' DataFrame to understand the available columns
crimen_past.columns

Index(['Location_Desc', 'Police_Beat', 'Street_Type', 'RecordID',
       'Date_Reported', 'Offense_Type', 'Block_Range', 'Street_Prefix',
       'Street_Name', 'Num_Offenses', 'Srce_YYMM', 'NGC', 'Match_addr', 'X',
       'Y', 'GEOID10', 'GEOID00', 'Police_Division', 'CTA_ID', 'CTA_Name',
       'Geocoded', 'Location_Code', 'Hour'],
      dtype='object')

In [65]:
# Dropping unnecessary columns from the 'crimen_past' DataFrame to streamline the dataset
crimen_past.drop(['Location_Desc','Police_Beat','Street_Type','RecordID',
                  'Date_Reported','Block_Range','Street_Prefix',
                  'Street_Name', 'Location_Code', 'Hour',
                  'Police_Division', 'NGC', 'Match_addr', 'GEOID10', 'GEOID00'], axis=1, inplace=True)

In [66]:
# Counting the occurrences of each unique value in the 'Offense_Type' column
conteo_valores = crimen_past['Offense_Type'].value_counts()

# Displaying the count of each unique value in the 'Offense_Type' column
print(conteo_valores)

Theft                 600981
Burglary              200450
Auto Theft            109185
Aggravated Assault     90075
Robbery                83792
Rape                    7622
Murder                  2088
1                         14
Name: Offense_Type, dtype: int64


In [67]:
# Define a function 'Condicion' that categorizes crime types as 'A' or 'B'
def Condicion(crimen_2014):
    if crimen_2014 in ['Aggravated Assault','Rape', 'Murder' ]:
        return 'A'
    else:
        return 'B'

# Apply the 'Condicion' function to create a new column 'A' based on the 'Offense_Type' column
crimen_past['A'] = crimen_past['Offense_Type'].apply(Condicion)

# Display the modified 'crimen_past' DataFrame with the new 'A' column
crimen_past

Offense_Type  Num_Offenses Srce_YYMM          X          Y CTA_ID  \
0           Burglary             1   2014_06 -95.333268  29.637514   K066   
1              Theft             1   2014_06 -95.331799  29.800081   K004   
2              Theft             1   2014_07 -95.516291  29.625560   K070   
3            Robbery             1   2014_07 -95.514200  29.736820   K080   
4           Burglary             1   2014_09 -95.195466  29.924672   K106   
...              ...           ...       ...        ...        ...    ...   
1094202     Burglary             1   2017_12 -95.208459  29.770717   K054   
1094203        Theft             1   2017_12 -95.218625  29.780813   K054   
1094204        Theft             1   2017_12 -95.290745  29.777468   K005   
1094205      Robbery             1   2017_12 -95.215481  29.772798   K054   
1094206        Theft             1   2017_12 -95.410353  29.690881   K019   

                             CTA_Name  Geocoded  A  
0                            Minnetex         1  B  
1                    Kashmere Gardens         1  B  
2                        Five Corners         1  B  
3                            Mid-West         1  B  
4                        Lake Houston         1  B  
...                               ...       ... ..  
1094202                    Northshore         1  B  
1094203                    Northshore         1  B  
1094204  Denver Harbor / Port Houston         1  B  
1094205                    Northshore         1  B  
1094206                     Astrodome         1  B  

[1094207 rows x 9 columns]

In [68]:
# Modify values in the 'Srce_YYMM' column by keeping only the first part before the underscore
crimen_past.Srce_YYMM = crimen_past.Srce_YYMM.apply(lambda cell: cell.split("_")[0])

# Group the DataFrame by specified columns, count occurrences, and reset the index
crimen_final_past = crimen_past.groupby(["Srce_YYMM", "CTA_ID", "CTA_Name", "Offense_Type", "A"]).count()["Num_Offenses"].reset_index()

# Rename columns 'CTA_ID' to 'cta_code' and 'CTA_Name' to 'cta_name' in the 'crimen_final_past' DataFrame
crimen_final_past.rename(columns={'CTA_ID': 'cta_code', 'CTA_Name': 'cta_name'}, inplace=True)

Now, in order to concatenate the two dataframes, it is essential to examine the columns of each dataframe. By comparing the column names, we can identify corresponding columns and ensure a seamless concatenation process. This step is crucial to maintain consistency and coherence in the combined dataset, allowing us to perform comprehensive analyses that span both datasets seamlessly.

In [69]:
crimen.columns

Index(['2019', 'cta_code', 'cta_name', 'Offense_Type', 'A', 'Num_Offenses'], dtype='object')

In [70]:
crimen_final_past.columns

Index(['Srce_YYMM', 'cta_code', 'cta_name', 'Offense_Type', 'A',
       'Num_Offenses'],
      dtype='object')

In [71]:
# Rename the column '2019' to 'Srce_YYMM' in the 'crimen' DataFrame
crimen.rename(columns={'2019': 'Srce_YYMM'}, inplace=True)

# Select specific columns in the 'crimen' DataFrame
crimen.loc[:,['Srce_YYMM', 'cta_code', 'Offense_Type', 'A', 'Num_Offenses']]

/var/folders/_n/vpsjjrqj2yb_yy6q_1xgvwg40000gn/T/ipykernel_2960/2429771425.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  crimen.rename(columns={'2019': 'Srce_YYMM'}, inplace=True)


Srce_YYMM cta_code                        Offense_Type  A  Num_Offenses
0           2019     K001  Murder & Nonnegligent Manslaughter  A             1
1           2019     K001                       Forcible Rape  A             2
2           2019     K001                     Forcible Sodomy  A             3
3           2019     K001                   Forcible Fondling  A             1
4           2019     K001                             Robbery  A            24
...          ...      ...                                 ... ..           ...
13529       2021     K140                  All Other Offenses  B             1
13530       2021     K143                      Simple Assault  A             1
13531       2021     K143                 Theft From Building  A             1
13532       2021     K143                   All Other Larceny  A             4
13533       2021     K143                  All Other Offenses  B             1

[13534 rows x 5 columns]

In [72]:
# Concatenate the two DataFrames 'crimen_final_past' and 'crimen' into a new DataFrame 'tabla_final'
tabla_final = pd.concat([crimen_final_past, crimen])

# Reset the index of the 'tabla_final' DataFrame
tabla_final.reset_index(inplace=True)

# Drop the 'index' column from the 'tabla_final' DataFrame
tabla_final.drop(['index'], axis=1, inplace=True)

# Display the final DataFrame
tabla_final

Srce_YYMM cta_code  cta_name         Offense_Type  A  Num_Offenses
0          2009     K001  Downtown   Aggravated Assault  A            65
1          2009     K001  Downtown           Auto Theft  B           137
2          2009     K001  Downtown             Burglary  B            48
3          2009     K001  Downtown               Murder  A             2
4          2009     K001  Downtown                 Rape  A             6
...         ...      ...       ...                  ... ..           ...
20145      2021     K140   Baytown   All Other Offenses  B             1
20146      2021     K143   Tomball       Simple Assault  A             1
20147      2021     K143   Tomball  Theft From Building  A             1
20148      2021     K143   Tomball    All Other Larceny  A             4
20149      2021     K143   Tomball   All Other Offenses  B             1

[20150 rows x 6 columns]

# Conclusion

In our data journey, we initiated by meticulously collecting crime data from cite sources, spanning the years 2009 to 2021, to construct a comprehensive understanding of Houston's crime landscape. The process involved adherence to stringent scrubbing procedures to ensure data accuracy and reliability. We employed strategic techniques such as handling missing values, column renaming, and selective column dropping to refine and prepare the datasets. As a result, our meticulously scrubbed and consolidated dataset, 'tabla_final,' now stands as a robust foundation for subsequent analyses, setting the stage for comprehensive insights into Houston's crime trends, geographical patterns, and offense classifications.